File: impl/python/src/img_drv/nix/lexer.py
"""The Nix lexer, built with PLY's ``lex``.

Generated rather than hand-written for the reason in
``docs/decisions/2026-08-02-nix-frontend-build-not-reuse.md``: Nix's own lexer
is Flex, its tokens overlap, and maximal munch resolves them. The sharpest case
is ``x:x``, which is a URI and NOT a lambda, because the URI rule matches more
characters than the identifier rule does.

ONE DIFFERENCE FROM FLEX THAT SHAPES THIS FILE
-----------------------------------------------

Flex and ocamllex pick the LONGEST match; PLY builds one Python regex by
alternation, and :mod:`re` picks the FIRST alternative that matches. So order
here is load-bearing where in ``lexer.mll`` it was not: ``uri`` must precede
``id``, ``path`` must precede both, and every multi-character operator must
precede its one-character prefix. PLY adds function rules in definition order
and string rules by decreasing pattern length, so the tricky ones are written
as functions to fix their order explicitly.

Interpolation needs lexer STATE, exactly as Nix's start conditions do. The
state stack is explicit rather than PLY's, because ``}`` is ambiguous without
it: it closes an attribute set in expression state and ends an antiquotation
otherwise.
"""

from __future__ import annotations

import re
from typing import Any

from ply import lex  # type: ignore[import-untyped]

__all__ = ["NixLexer", "tokens"]

reserved = {
    "if": "IF",
    "then": "THEN",
    "else": "ELSE",
    "assert": "ASSERT",
    "with": "WITH",
    "let": "LET",
    "in": "IN",
    "rec": "REC",
    "inherit": "INHERIT",
    "or": "OR_KW",
}

tokens = [
    "INT",
    "FLOAT",
    "ID",
    "STR",
    "ESTR",
    "PATH",
    "SPATH",
    "URI",
    "PATH_START",
    "PATH_STR",
    "PATH_END",
    "DQUOTE",
    "IND_OPEN",
    "IND_CLOSE",
    "DOLLAR_CURLY",
    "LCURLY",
    "RCURLY",
    "LPAREN",
    "RPAREN",
    "LBRACK",
    "RBRACK",
    "SEMI",
    "COMMA",
    "COLON",
    "AT",
    "DOT",
    "ELLIPSIS",
    "ASSIGN",
    "QUESTION",
    "EQ",
    "NEQ",
    "LEQ",
    "GEQ",
    "LT",
    "GT",
    "AND",
    "OR",
    "IMPL",
    "UPDATE",
    "CONCAT",
    "PLUS",
    "MINUS",
    "TIMES",
    "SLASH",
    "NOT",
    *reserved.values(),
]

states = (
    ("str", "exclusive"),
    ("indstr", "exclusive"),
)

_PATH_CHAR = r"[A-Za-z0-9._+\-]"
#: The literal segment that can follow an interpolation inside a path.
_PATH_TAIL = re.compile(rf"(?:{_PATH_CHAR}|/)*")

_ESCAPES = {"n": "\n", "r": "\r", "t": "\t"}


def _unescape(c: str) -> str:
    return _ESCAPES.get(c, c)


def _unescape_run(text: str) -> str:
    """Resolve the backslash escapes inside a matched run.

    Nix's string rule matches a MAXIMAL run that already contains its escapes,
    so unescaping happens on the whole token rather than one escape at a time.
    Reproducing that boundary matters: string parts are never merged, so where
    the lexer splits is exactly what the printed tree shows.
    """
    out: list[str] = []
    i = 0
    while i < len(text):
        if text[i] == "\\" and i + 1 < len(text):
            out.append(_unescape(text[i + 1]))
            i += 2
        else:
            out.append(text[i])
            i += 1
    return "".join(out)


# INITIAL state


def t_ignore_comment_line(t: Any) -> None:
    r"\#[^\n]*"


def t_ignore_comment_block(t: Any) -> None:
    r"/\*(?:.|\n)*?\*/"
    t.lexer.lineno += t.value.count("\n")


def t_newline(t: Any) -> None:
    r"\n+"
    t.lexer.lineno += len(t.value)


t_ignore = " \t\r"


def t_URI(t: Any) -> Any:
    r"[A-Za-z][A-Za-z0-9+\-.]*:[A-Za-z0-9%/?:@&=+$,_.!~*'\-]+"
    return t


def t_PATH_START(t: Any) -> Any:
    r"(?:~(?:/[A-Za-z0-9._+\-]*)+|[A-Za-z0-9._+\-]*(?:/[A-Za-z0-9._+\-]*)+)\$\{"
    # A path containing an interpolation. The prefix is looser than `path`: it
    # may end in a bare slash, as in `./${v}`.
    t.value = t.value[:-2]
    t.lexer.nix_push("path")
    t.lexer.nix_push("expr")
    return t


def t_PATH(t: Any) -> Any:
    r"~(?:/[A-Za-z0-9._+\-]+)+/?|[A-Za-z0-9._+\-]*(?:/[A-Za-z0-9._+\-]+)+/?"
    return t


def t_SPATH(t: Any) -> Any:
    r"<[A-Za-z0-9._+\-]+(?:/[A-Za-z0-9._+\-]+)*>"
    t.value = t.value[1:-1]
    return t


def t_FLOAT(t: Any) -> Any:
    r"(?:[0-9]+\.[0-9]*|\.[0-9]+)(?:[Ee][+\-]?[0-9]+)?"
    t.value = float(t.value)
    return t


def t_INT(t: Any) -> Any:
    r"[0-9]+"
    t.value = int(t.value)
    return t


def t_ID(t: Any) -> Any:
    r"[A-Za-z_][A-Za-z0-9_'\-]*"
    t.type = reserved.get(t.value, "ID")
    return t


def t_IND_OPEN(t: Any) -> Any:
    r"''[ ]*\n|''"
    # The opening delimiter swallows any spaces and ONE newline after it, which
    # is why an indented string on its own line has no leading blank line.
    t.lexer.lineno += t.value.count("\n")
    t.lexer.nix_push("indstr")
    return t


def t_DQUOTE(t: Any) -> Any:
    r"\""
    t.lexer.nix_push("str")
    return t


def t_DOLLAR_CURLY(t: Any) -> Any:
    r"\$\{"
    t.lexer.nix_push("expr")
    return t


def t_LCURLY(t: Any) -> Any:
    r"\{"
    t.lexer.nix_push("expr")
    return t


def t_RCURLY(t: Any) -> Any:
    r"\}"
    t.lexer.nix_pop()
    return t


def t_ELLIPSIS(t: Any) -> Any:
    r"\.\.\."
    return t


def t_IMPL(t: Any) -> Any:
    r"->"
    return t


def t_UPDATE(t: Any) -> Any:
    r"//"
    return t


def t_CONCAT(t: Any) -> Any:
    r"\+\+"
    return t


def t_EQ(t: Any) -> Any:
    r"=="
    return t


def t_NEQ(t: Any) -> Any:
    r"!="
    return t


def t_LEQ(t: Any) -> Any:
    r"<="
    return t


def t_GEQ(t: Any) -> Any:
    r">="
    return t


def t_AND(t: Any) -> Any:
    r"&&"
    return t


def t_OR(t: Any) -> Any:
    r"\|\|"
    return t


t_LPAREN = r"\("
t_RPAREN = r"\)"
t_LBRACK = r"\["
t_RBRACK = r"\]"
t_SEMI = r";"
t_COMMA = r","
t_COLON = r":"
t_AT = r"@"
t_DOT = r"\."
t_ASSIGN = r"="
t_QUESTION = r"\?"
t_LT = r"<"
t_GT = r">"
t_PLUS = r"\+"
t_MINUS = r"-"
t_TIMES = r"\*"
t_SLASH = r"/"
t_NOT = r"!"


def t_error(t: Any) -> None:
    raise SyntaxError(
        f"{t.lexer.lineno}:{_column(t)}: unexpected character {t.value[0]!r}"
    )


# Double-quoted string state.
#
# The chunk boundaries MATTER, because nothing merges them afterwards: what the
# lexer splits is exactly what the printed tree shows. So the runs below are
# transcribed from NixOS/nix lexer.l rather than invented.

t_str_ignore = ""


def t_str_DOLLAR_CURLY(t: Any) -> Any:
    r"\$\{"
    t.lexer.nix_push("expr")
    return t


def t_str_trailing(t: Any) -> Any:
    r"(?:[^$\"\\]|\$[^{\"\\]|\\[\s\S]|\$\\[\s\S])*\$(?=\")"
    # Nix's FIRST string rule uses flex TRAILING CONTEXT: a run ending in a
    # dollar that is followed by the closing quote. That dollar cannot be
    # absorbed by the ordinary run below, which needs a character after it, so
    # without this rule a string ending in a dollar becomes a two-part
    # concatenation. Python's lookahead expresses the trailing context directly.
    t.type = "STR"
    t.lexer.lineno += t.value.count("\n")
    t.value = _unescape_run(t.value)
    return t


def t_str_run(t: Any) -> Any:
    r"(?:[^$\"\\]|\$[^{\"\\]|\\[\s\S]|\$\\[\s\S])+"
    t.type = "STR"
    t.lexer.lineno += t.value.count("\n")
    t.value = _unescape_run(t.value)
    return t


def t_str_DQUOTE(t: Any) -> Any:
    r"\""
    t.lexer.nix_pop()
    return t


def t_str_fallback(t: Any) -> Any:
    r"\$\\|\$|\\"
    t.type = "STR"
    return t


def t_str_error(t: Any) -> None:
    raise SyntaxError(f"{t.lexer.lineno}: unterminated string")


# Indented string state.

t_indstr_ignore = ""


def t_indstr_run(t: Any) -> Any:
    r"(?:[^$']|\$[^{']|'[^'$])+"
    # One MAXIMAL run. Note what it EXCLUDES: a quote before another quote or
    # before a dollar, so the two-quote escapes below still get their chance.
    t.type = "STR"
    t.lexer.lineno += t.value.count("\n")
    return t


def t_indstr_esc_quotes(t: Any) -> Any:
    r"'''"
    t.type = "ESTR"
    t.value = "''"
    return t


def t_indstr_esc_dollar(t: Any) -> Any:
    r"''\$"
    t.type = "ESTR"
    t.value = "$"
    return t


def t_indstr_esc_backslash(t: Any) -> Any:
    r"''\\[\s\S]"
    t.type = "ESTR"
    t.value = _unescape(t.value[3])
    return t


def t_indstr_IND_CLOSE(t: Any) -> Any:
    r"''"
    t.lexer.nix_pop()
    return t


def t_indstr_DOLLAR_CURLY(t: Any) -> Any:
    r"\$\{"
    t.lexer.nix_push("expr")
    return t


def t_indstr_fallback(t: Any) -> Any:
    r"'|\$"
    # Every chunk that is not the plain run takes no part in indentation, which
    # is what Nix's StringToken.hasIndentation records. An escaped newline
    # produces a REAL newline character, so if the dedent pass scanned it the
    # text after would look like an unindented line and the whole string would
    # stop being dedented.
    t.type = "ESTR"
    return t


def t_indstr_error(t: Any) -> None:
    raise SyntaxError(f"{t.lexer.lineno}: unterminated indented string")


def _column(t: Any) -> int:
    start: int = t.lexer.lexdata.rfind("\n", 0, t.lexpos) + 1
    pos: int = t.lexpos
    return pos - start + 1


class NixLexer:
    """A stateful wrapper around PLY's lexer.

    Two things PLY does not do on its own live here.

    The STATE STACK, because ``}`` is ambiguous: it closes an attribute set in
    expression state and ends an antiquotation otherwise, so the lexer has to
    remember which.

    The PATH TAIL. After an interpolation inside a path closes, the remaining
    path characters are consumed here and a ``PATH_END`` is queued, because
    that end has to be signalled WITHOUT consuming the character that caused
    it and PLY rejects a zero-length match. A one-token queue is the smallest
    thing that expresses it.
    """

    def __init__(self) -> None:
        # Cloned rather than rebuilt: PLY compiles one master regex per lexer,
        # and the corpus check builds thousands of these.
        self._lexer = _MASTER.clone()
        self._lexer.nix_push = self._push
        self._lexer.nix_pop = self._pop
        self._queue: list[Any] = []
        self._stack: list[str] = ["expr"]

    def _begin(self, state: str) -> None:
        self._lexer.begin("INITIAL" if state in ("expr", "path") else state)

    def _push(self, state: str) -> None:
        self._stack.append(state)
        self._begin(state)

    def _pop(self) -> str:
        top = self._stack.pop() if len(self._stack) > 1 else "expr"
        self._begin(self._stack[-1])
        if self._stack[-1] == "path":
            self._scan_path_tail()
        return top

    def _scan_path_tail(self) -> None:
        """Consume the literal segment after an interpolation inside a path."""
        lexer = self._lexer
        m = _PATH_TAIL.match(lexer.lexdata, lexer.lexpos)
        segment = m.group(0) if m else ""
        lexer.lexpos += len(segment)
        if segment:
            self._queue.append(_tok("PATH_STR", segment, lexer.lineno))
        if lexer.lexdata[lexer.lexpos : lexer.lexpos + 2] == "${":
            lexer.lexpos += 2
            self._queue.append(_tok("DOLLAR_CURLY", "${", lexer.lineno))
            self._stack.append("expr")
        else:
            self._queue.append(_tok("PATH_END", "", lexer.lineno))
            self._stack.pop()
            self._begin(self._stack[-1])

    def input(self, text: str) -> None:
        self._lexer.input(text)
        self._lexer.lineno = 1
        self._stack = ["expr"]
        self._queue = []
        self._begin("expr")

    def token(self) -> Any:
        if self._queue:
            return self._queue.pop(0)
        tok: Any = self._lexer.token()
        return tok


def _tok(type_: str, value: str, lineno: int) -> Any:
    t = lex.LexToken()
    t.type = type_
    t.value = value
    t.lineno = lineno
    t.lexpos = 0
    return t


#: Built once; every :class:`NixLexer` clones it.
_MASTER = lex.lex(reflags=re.VERBOSE)

File: impl/rust/src/nix/lexer.rs
//! The Nix lexer, built on `logos`.
//!
//! Generated rather than hand-written for the reason in
//! `docs/decisions/2026-08-02-nix-frontend-build-not-reuse.md`: Nix's own lexer
//! is Flex, its tokens overlap, and maximal munch resolves them. The sharpest
//! case is `x:x`, which is a URI and NOT a lambda, because the URI rule matches
//! more characters than the identifier rule does. `logos` picks the longest
//! match too, so that falls out rather than needing rule ordering the way PLY
//! does.
//!
//! # Modes, and how they are done here
//!
//! Interpolation needs lexer STATE, exactly as Nix's start conditions do:
//! inside a string `${` switches back to expression lexing, and the matching
//! `}` has to be told apart from an attribute set's closing brace.
//!
//! `logos` has no start conditions, so each mode is its own token enum. Rather
//! than `morph` between them, which moves the lexer, this keeps a plain
//! `(source, position)` pair and builds a fresh `logos::Lexer` over the
//! remaining slice for each token. Constructing one is O(1), and it makes the
//! mode a value in a stack instead of a type-level state the borrow checker
//! has to be talked through.

use logos::Logos;

/// A token, in the shape LALRPOP wants: `(start, token, end)`.
pub type Spanned = (usize, Tok, usize);

/// What the lexer is in the middle of.
///
/// `Path` is here because an interpolated path continues after its
/// interpolation closes, so the lexer has to know to keep reading path
/// characters rather than expression tokens.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Mode {
    Expr,
    Str,
    IndStr,
    Path,
}

/// The tokens the grammar consumes.
///
/// One enum for every mode, because the grammar does not care which mode
/// produced a chunk, only what it is.
///
/// `missing_docs` is allowed HERE and nowhere else in the crate: forty of these
/// are punctuation whose name is its whole meaning, and a doc comment saying
/// "a semicolon" on `Semi` is noise that hides the five variants that DO carry
/// a rule. Those five are documented.
#[derive(Debug, Clone, PartialEq)]
#[allow(missing_docs)]
pub enum Tok {
    Int(i64),
    Float(f64),
    Id(String),
    /// A string chunk that DOES take part in an indented string's dedenting.
    Str(String),
    /// A chunk produced by an escape, which does NOT.
    ///
    /// This is Nix's `StringToken.hasIndentation`, and it matters for a reason
    /// that is easy to miss: an escaped newline is a real newline character,
    /// so scanning it would make the following text look like the start of an
    /// unindented line and switch the dedent off for the whole string.
    EStr(String),
    Path(String),
    SPath(String),
    Uri(String),
    /// The literal prefix of an interpolated path, `${` already consumed.
    PathStart(String),
    /// A literal segment BETWEEN interpolations in a path.
    PathStr(String),
    /// The end of an interpolated path. Emitted without consuming input.
    PathEnd,
    If,
    Then,
    Else,
    Assert,
    With,
    Let,
    In,
    Rec,
    Inherit,
    OrKw,
    DQuote,
    IndOpen,
    IndClose,
    DollarCurly,
    LCurly,
    RCurly,
    LParen,
    RParen,
    LBrack,
    RBrack,
    Semi,
    Comma,
    Colon,
    At,
    Dot,
    Ellipsis,
    Assign,
    Question,
    Eq,
    Neq,
    Leq,
    Geq,
    Lt,
    Gt,
    And,
    Or,
    Impl,
    Update,
    Concat,
    Plus,
    Minus,
    Times,
    Slash,
    Not,
}

/// A lexical error, carrying the byte offset so a message can be located.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LexError {
    /// Byte offset into the source where the lexer gave up.
    pub offset: usize,
    /// What went wrong, in the same shape as the other implementations'.
    pub message: String,
}

impl std::fmt::Display for Tok {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // Only ever used inside a parse-error message, so the debug shape is
        // the useful one: it names the variant AND its payload.
        write!(f, "{self:?}")
    }
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "offset {}: {}", self.offset, self.message)
    }
}

#[derive(Logos, Debug, PartialEq)]
#[logos(skip r"[ \t\r\n]+")]
// `allow_greedy` because a line comment genuinely runs to the newline;
// logos warns since such a pattern can scan far, which is what it must do.
#[logos(skip(r"#[^\n]*", allow_greedy = true))]
#[logos(skip r"/\*([^*]|\*+[^*/])*\*+/")]
enum ExprTok {
    // `uri` before `id` is not needed here: logos takes the longest match, so
    // `x:x` lexes as a URI on its own.
    #[regex(r"[A-Za-z][A-Za-z0-9+\-.]*:[A-Za-z0-9%/?:@&=+$,_.!~*'\-]+")]
    Uri,
    #[token("if")]
    If,
    #[token("then")]
    Then,
    #[token("else")]
    Else,
    #[token("assert")]
    Assert,
    #[token("with")]
    With,
    #[token("let")]
    Let,
    #[token("in")]
    In,
    #[token("rec")]
    Rec,
    #[token("inherit")]
    Inherit,
    #[token("or")]
    OrKw,
    #[regex(r"[A-Za-z_][A-Za-z0-9_'\-]*")]
    Id,
    #[regex(r"([0-9]+\.[0-9]*|\.[0-9]+)([Ee][+\-]?[0-9]+)?")]
    Float,
    #[regex(r"[0-9]+")]
    Int,
    // A path containing an interpolation. The prefix is looser than `path`: it
    // may end in a bare slash, as in `./${v}`.
    #[regex(r"(~(/[A-Za-z0-9._+\-]*)+|[A-Za-z0-9._+\-]*(/[A-Za-z0-9._+\-]*)+)\$\{")]
    PathStart,
    #[regex(r"~(/[A-Za-z0-9._+\-]+)+/?")]
    #[regex(r"[A-Za-z0-9._+\-]*(/[A-Za-z0-9._+\-]+)+/?")]
    Path,
    #[regex(r"<[A-Za-z0-9._+\-]+(/[A-Za-z0-9._+\-]+)*>")]
    SPath,
    // The opening delimiter swallows any spaces and ONE newline after it, which
    // is why an indented string on its own line has no leading blank line.
    #[regex(r"''[ ]*\n")]
    #[token("''")]
    IndOpen,
    #[token("\"")]
    DQuote,
    #[token("${")]
    DollarCurly,
    #[token("{")]
    LCurly,
    #[token("}")]
    RCurly,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("[")]
    LBrack,
    #[token("]")]
    RBrack,
    #[token(";")]
    Semi,
    #[token(",")]
    Comma,
    #[token(":")]
    Colon,
    #[token("@")]
    At,
    #[token("...")]
    Ellipsis,
    #[token(".")]
    Dot,
    #[token("=")]
    Assign,
    #[token("?")]
    Question,
    #[token("==")]
    Eq,
    #[token("!=")]
    Neq,
    #[token("<=")]
    Leq,
    #[token(">=")]
    Geq,
    #[token("<")]
    Lt,
    #[token(">")]
    Gt,
    #[token("&&")]
    And,
    #[token("||")]
    Or,
    #[token("->")]
    Impl,
    #[token("//")]
    Update,
    #[token("++")]
    Concat,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Times,
    #[token("/")]
    Slash,
    #[token("!")]
    Not,
}

/// Inside a double-quoted string.
///
/// The chunk boundaries MATTER, because nothing merges them afterwards: what
/// the lexer splits is exactly what the printed tree shows. So the runs are
/// transcribed from `NixOS/nix` `lexer.l` rather than invented.
#[derive(Logos, Debug, PartialEq)]
enum StrTok {
    #[token("\"")]
    DQuote,
    #[token("${")]
    DollarCurly,
    // Nix's FIRST rule uses flex TRAILING CONTEXT: a run ending in a dollar
    // that is followed by the closing quote. That dollar cannot be absorbed by
    // the ordinary run, which needs a character after it, so without this a
    // string ending in a dollar becomes a two-part concatenation. `logos` has
    // no lookahead, so the quote is matched and pushed back by the wrapper.
    #[regex(r#"([^$"\\]|\$[^{"\\]|\\[\s\S]|\$\\[\s\S])*\$""#)]
    RunThenQuote,
    #[regex(r#"([^$"\\]|\$[^{"\\]|\\[\s\S]|\$\\[\s\S])+"#)]
    Run,
    #[regex(r"\$\\|\$|\\")]
    Fallback,
}

/// Inside an indented string.
#[derive(Logos, Debug, PartialEq)]
enum IndTok {
    // One MAXIMAL run. Note what it EXCLUDES: a quote before another quote or
    // before a dollar, so the two-quote escapes below still get their chance.
    #[regex(r"([^$']|\$[^{']|'[^'$])+")]
    Run,
    #[token("'''")]
    EscQuotes,
    #[token("''$")]
    EscDollar,
    #[regex(r"''\\[\s\S]")]
    EscBackslash,
    #[token("''")]
    IndClose,
    #[token("${")]
    DollarCurly,
    #[token("'")]
    Quote,
    #[token("$")]
    Dollar,
}

fn unescape(c: char) -> char {
    match c {
        'n' => '\n',
        'r' => '\r',
        't' => '\t',
        c => c,
    }
}

/// Resolve the backslash escapes inside a matched run.
///
/// Nix's string rule matches a MAXIMAL run that already contains its escapes,
/// so unescaping happens on the whole token rather than one escape at a time.
fn unescape_run(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    let mut chars = text.chars();
    while let Some(c) = chars.next() {
        if c == '\\' {
            match chars.next() {
                Some(next) => out.push(unescape(next)),
                None => out.push(c),
            }
        } else {
            out.push(c);
        }
    }
    out
}

/// The token stream LALRPOP drives.
pub struct Lexer<'a> {
    src: &'a str,
    pos: usize,
    stack: Vec<Mode>,
    queue: std::collections::VecDeque<Spanned>,
}

impl<'a> Lexer<'a> {
    /// Start lexing `src` in expression mode.
    pub fn new(src: &'a str) -> Self {
        Lexer {
            src,
            pos: 0,
            stack: vec![Mode::Expr],
            queue: std::collections::VecDeque::new(),
        }
    }

    fn mode(&self) -> Mode {
        *self.stack.last().unwrap_or(&Mode::Expr)
    }

    fn push(&mut self, mode: Mode) {
        self.stack.push(mode);
    }

    fn pop(&mut self) {
        if self.stack.len() > 1 {
            self.stack.pop();
        }
    }

    /// Consume the literal segment after an interpolation inside a path.
    ///
    /// The end of a path has to be signalled WITHOUT consuming the character
    /// that ended it, so it is queued here rather than matched by a rule.
    fn scan_path_tail(&mut self) {
        let rest = &self.src[self.pos..];
        let len = rest
            .find(|c: char| !(c.is_ascii_alphanumeric() || "._+-/".contains(c)))
            .unwrap_or(rest.len());
        let start = self.pos;
        if len > 0 {
            self.queue
                .push_back((start, Tok::PathStr(rest[..len].to_string()), start + len));
            self.pos += len;
        }
        if self.src[self.pos..].starts_with("${") {
            let at = self.pos;
            self.pos += 2;
            self.queue.push_back((at, Tok::DollarCurly, at + 2));
            self.push(Mode::Expr);
        } else {
            self.queue.push_back((self.pos, Tok::PathEnd, self.pos));
            self.pop();
        }
    }

    fn err(&self, message: &str) -> LexError {
        LexError {
            offset: self.pos,
            message: message.to_string(),
        }
    }

    fn next_expr(&mut self) -> Option<Result<Spanned, LexError>> {
        let rest = &self.src[self.pos..];
        let mut lex = ExprTok::lexer(rest);
        let token = lex.next()?;
        let span = lex.span();
        let (start, end) = (self.pos + span.start, self.pos + span.end);
        let text = &rest[span.clone()];
        self.pos = end;
        let tok = match token {
            Err(()) => return Some(Err(self.err("unexpected character"))),
            Ok(t) => t,
        };
        let out = match tok {
            ExprTok::Uri => Tok::Uri(text.to_string()),
            ExprTok::If => Tok::If,
            ExprTok::Then => Tok::Then,
            ExprTok::Else => Tok::Else,
            ExprTok::Assert => Tok::Assert,
            ExprTok::With => Tok::With,
            ExprTok::Let => Tok::Let,
            ExprTok::In => Tok::In,
            ExprTok::Rec => Tok::Rec,
            ExprTok::Inherit => Tok::Inherit,
            ExprTok::OrKw => Tok::OrKw,
            ExprTok::Id => Tok::Id(text.to_string()),
            ExprTok::Float => match text.parse() {
                Ok(f) => Tok::Float(f),
                Err(_) => return Some(Err(self.err("bad float"))),
            },
            ExprTok::Int => match text.parse() {
                Ok(n) => Tok::Int(n),
                Err(_) => return Some(Err(self.err("bad integer"))),
            },
            ExprTok::PathStart => {
                self.push(Mode::Path);
                self.push(Mode::Expr);
                Tok::PathStart(text[..text.len() - 2].to_string())
            }
            ExprTok::Path => Tok::Path(text.to_string()),
            ExprTok::SPath => Tok::SPath(text[1..text.len() - 1].to_string()),
            ExprTok::IndOpen => {
                self.push(Mode::IndStr);
                Tok::IndOpen
            }
            ExprTok::DQuote => {
                self.push(Mode::Str);
                Tok::DQuote
            }
            ExprTok::DollarCurly => {
                self.push(Mode::Expr);
                Tok::DollarCurly
            }
            ExprTok::LCurly => {
                self.push(Mode::Expr);
                Tok::LCurly
            }
            ExprTok::RCurly => {
                self.pop();
                if self.mode() == Mode::Path {
                    self.scan_path_tail();
                }
                Tok::RCurly
            }
            ExprTok::LParen => Tok::LParen,
            ExprTok::RParen => Tok::RParen,
            ExprTok::LBrack => Tok::LBrack,
            ExprTok::RBrack => Tok::RBrack,
            ExprTok::Semi => Tok::Semi,
            ExprTok::Comma => Tok::Comma,
            ExprTok::Colon => Tok::Colon,
            ExprTok::At => Tok::At,
            ExprTok::Ellipsis => Tok::Ellipsis,
            ExprTok::Dot => Tok::Dot,
            ExprTok::Assign => Tok::Assign,
            ExprTok::Question => Tok::Question,
            ExprTok::Eq => Tok::Eq,
            ExprTok::Neq => Tok::Neq,
            ExprTok::Leq => Tok::Leq,
            ExprTok::Geq => Tok::Geq,
            ExprTok::Lt => Tok::Lt,
            ExprTok::Gt => Tok::Gt,
            ExprTok::And => Tok::And,
            ExprTok::Or => Tok::Or,
            ExprTok::Impl => Tok::Impl,
            ExprTok::Update => Tok::Update,
            ExprTok::Concat => Tok::Concat,
            ExprTok::Plus => Tok::Plus,
            ExprTok::Minus => Tok::Minus,
            ExprTok::Times => Tok::Times,
            ExprTok::Slash => Tok::Slash,
            ExprTok::Not => Tok::Not,
        };
        Some(Ok((start, out, end)))
    }

    fn next_str(&mut self) -> Option<Result<Spanned, LexError>> {
        let rest = &self.src[self.pos..];
        let mut lex = StrTok::lexer(rest);
        let token = lex.next()?;
        let span = lex.span();
        let start = self.pos + span.start;
        let mut end = self.pos + span.end;
        let text = &rest[span.clone()];
        let tok = match token {
            Err(()) => return Some(Err(self.err("unterminated string"))),
            Ok(StrTok::DQuote) => {
                self.pop();
                Tok::DQuote
            }
            Ok(StrTok::DollarCurly) => {
                self.push(Mode::Expr);
                Tok::DollarCurly
            }
            Ok(StrTok::RunThenQuote) => {
                // Push the closing quote back, which is what flex's trailing
                // context does for free.
                end -= 1;
                Tok::Str(unescape_run(&text[..text.len() - 1]))
            }
            Ok(StrTok::Run) => Tok::Str(unescape_run(text)),
            Ok(StrTok::Fallback) => Tok::Str(text.to_string()),
        };
        self.pos = end;
        Some(Ok((start, tok, end)))
    }

    fn next_ind(&mut self) -> Option<Result<Spanned, LexError>> {
        let rest = &self.src[self.pos..];
        let mut lex = IndTok::lexer(rest);
        let token = lex.next()?;
        let span = lex.span();
        let (start, end) = (self.pos + span.start, self.pos + span.end);
        let text = &rest[span.clone()];
        self.pos = end;
        let tok = match token {
            Err(()) => {
                return Some(Err(self.err("unterminated indented string")));
            }
            Ok(IndTok::Run) => Tok::Str(text.to_string()),
            Ok(IndTok::EscQuotes) => Tok::EStr("''".to_string()),
            Ok(IndTok::EscDollar) => Tok::EStr("$".to_string()),
            Ok(IndTok::EscBackslash) => {
                let c = text.chars().nth(3).unwrap_or('\\');
                Tok::EStr(unescape(c).to_string())
            }
            Ok(IndTok::IndClose) => {
                self.pop();
                Tok::IndClose
            }
            Ok(IndTok::DollarCurly) => {
                self.push(Mode::Expr);
                Tok::DollarCurly
            }
            Ok(IndTok::Quote) => Tok::EStr("'".to_string()),
            Ok(IndTok::Dollar) => Tok::EStr("$".to_string()),
        };
        Some(Ok((start, tok, end)))
    }
}

impl Iterator for Lexer<'_> {
    type Item = Result<Spanned, LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(queued) = self.queue.pop_front() {
            return Some(Ok(queued));
        }
        match self.mode() {
            Mode::Expr | Mode::Path => self.next_expr(),
            Mode::Str => self.next_str(),
            Mode::IndStr => self.next_ind(),
        }
    }
}
